### PR TITLE
Tables: Pills for Table Cells

### PR DESCRIPTION
**Pull request by timlevett**
After reviewing ideas from the [first PR](https://github.com/grafana/grafana/pull/107354) and the refactor, I decided to revamp this a bit.

image

Let me know what you think. 

https://github.com/user-attachments/assets/52c696f6-608e-48e8-b445-46312fa0e561

**What is this feature?**

Pills for tables for either json array of strings or csv of strings.

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

Pills are a nice way to visualize a small list of strings or a json array of strings.

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

Anyone who loves table data

**Which issue(s) does this PR fix?**:

This is v2 of https://github.com/grafana/grafana/pull/107354

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by timlevett** [comment]
What's new draft; https://admin.grafana.com/content-admin/#/collections/whats-new/entries/2025-07-02-pills-for-table-cells

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+function getPillColor(pill: string, cellOptions: TableCellRendererProps['cellOptions']): string {
+  if (!isPillCellOptions(cellOptions)) {
+    return getDeterministicColor(pill);
+  }
+
+  const colorMode = cellOptions.colorMode || 'auto';
+
+  // Fixed color mode (highest priority)
+  if (colorMode === 'fixed' && cellOptions.color) {
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx:117]: how useful is it to have all pills in the column be the same color?

we should support value mappings here. possibly even something like key-value pairs with mappings based on key rather than value

Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx:117]: yeah, fixed was an easy one to add, all the components where there so it was an easy win, will it be used, we could track it but its not that much weight really. 

I'm happy to try to add in kv mapping, maybe in next iteration? I did have it at first, but it made it a bit more complicated. I didn't try with kv pairs tho, just a mapping similar to our other kv mapping we have.

What do you think @leeoniya , do we need kv pairs to move forward or is this a good first iteration?

Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx:117]: Added in the latest commits, have a look @leeoniya 😄 , added video to the main PR description.

File: packages/grafana-schema/src/common/common.gen.ts [hunk 2]
```diff
@@ -838,7 +839,38 @@ export enum TableCellHeight {
...
+
+/**
+ * Field options for each field within a table (e.g 10, "The String", 64.20, etc.)
+ * Generally defines alignment, filtering capabilties, display options, etc.
+ */
+export interface TableFieldOptions {
+  align: FieldTextAlignment;
+  cellOptions: TableCellOptions;
+  /**
+   * This field is deprecated in favor of using cellOptions
```
Comment (jesdavpet) [inline packages/grafana-schema/src/common/common.gen.ts:852]: **suggestion(non-blocking):** Could be nice to use the TS+JSDoc `@deprecated` annotation here if we can? _(But I'm not quite sure if this file is generated or user editable based on `.gen` in the filename.)_

- https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated

```suggestion
   * @deprecated This field is deprecated in favor of using cellOptions
```

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx
Comment (jesdavpet) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx:1]: 🏆 Thank you for adding these tests !!!

**Review by jesdavpet — APPROVED** [review]
Thanks for adding this feature @timlevett ! 🎨 ✨ I've QA'ed these changes locally with the `tableNextGen` feature toggle enabled, and everything looks good so far.

Screenshot 2025-07-04 at 6 07 40 PM

**suggestion(non-blocking):** One nice-to-have to consider including with this PR would be an minimal example of Table + Pills in a `gdev` dashboard panel. But I leave that to your discretion.

IMO this one's ready to go once you and @leeoniya arrive at a consensus on whether key/value mappings can be saved for a later iteration.

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx (hunk 1 above)
Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx:117]: We have this now, thanks for the suggestion ⭐

File: packages/grafana-schema/src/common/common.gen.ts (hunk 2 above)
Comment (timlevett) [reply packages/grafana-schema/src/common/common.gen.ts:852]: I think this is just generated code, i didn't add this for this PR, its just in a slightly different spot. if you see about line 947

Comment (fastfrwrd) [reply packages/grafana-schema/src/common/common.gen.ts:852]: yeah, we'd have to figure out if there was a way to do it with [cuetsy](https://github.com/grafana/cuetsy). I plan to try to get more in the weeds with cuetsy sometime soon, so we can table this discussion for that.

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx
Comment (fastfrwrd) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx:1]: +1!

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  });
+
+  it('should render pills for JSON array values', () => {
+    render(<PillCell {...defaultProps} value='["item1","item2","item3"]' />);
+    expect(screen.getByText('item1')).toBeInTheDocument();
+    expect(screen.getByText('item2')).toBeInTheDocument();
+    expect(screen.getByText('item3')).toBeInTheDocument();
+  });
+
+  it('should use mapped colors when colorMode is mapped', () => {
```
Comment (fastfrwrd) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx]: can I request from an organization perspective: I love that these tests exist for the color stuff because it exercises the code and ensures nothing throws or anything.

- can we group all these color tests in their own `describe` (`describe('color mapping', () => {});` or something)
- can we have a comment at the top explaining that the purpose of these tests is primarily to make sure the color logic executes without throwing, and that we'd need visual regression tests to truly test color.

just for posterity's sake, so people don't see the titles of the tests and misunderstand what they do and don't check.

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import { GrafanaTheme2, isDataFrame, classicColors, colorManipulator, Field } from '@grafana/data';
+import { TablePillCellOptions } from '@grafana/schema';
+
+import { useStyles2 } from '../../../../themes/ThemeContext';
+import { TableCellRendererProps } from '../types';
+
+export function PillCell({ value, field, justifyContent, cellOptions }: TableCellRendererProps) {
+  const styles = useStyles2(getStyles, justifyContent);
+
+  const pills = useMemo(() => {
```
Comment (fastfrwrd) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx]: **non-blocking, fine for a follow-up:** I would consider pulling the logic in here into its own method, exported from this file, called `inferPills` or something. It could be exported and tested separately, which would help us add a bunch more tests if edge cases come up that we want to support (feels inevitable with this one).

```ts
const pills = useMemo(() => inferPills(value), [value]);
```

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  if (pills.length === 0) {
+    return <div className={styles.cell}>-</div>;
+  }
+
+  return (
+    <div className={styles.cell}>
+      <div className={styles.pillsContainer}>
+        {pills.map((pill, index) => {
+          const bgColor = getPillColor(pill, cellOptions, field);
+          const textColor = colorManipulator.getContrastRatio('#FFFFFF', bgColor) >= 4.5 ? '#FFFFFF' : '#000000';
```
Comment (fastfrwrd) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx]: **please fix:** I think we should memoize this calc. we could have a `useMemo` above right after the step which builds the values which maps `string[]` to `Pill[]`, and then this map is just responsible for rendering.

Can I suggest this structure for Pill:

```ts
interface Pill {
  value: string;
  key: string; // `${pill}-${index}`
  bgColor: string;
  color: string;
}
```

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    // Check if field has value mappings
+    if (field.config.mappings && field.config.mappings.length > 0) {
+      // Use the field's display processor to get the mapped value
+      const displayValue = field.display!(pill);
+      if (displayValue.color) {
+        return displayValue.color;
+      }
+    }
+    // Fallback to default color for unmapped values
+    return cellOptions.color || '#FF780A';
```
Comment (fastfrwrd) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx]: **nit:** let's throw this color (used a couple lines later too) into a const.

```ts
const DEFAULT_PILL_BG_COLOR = '#FF780A';
```

File: packages/grafana-ui/src/components/Table/utils.ts [hunk 7]
```diff
@@ -196,6 +196,8 @@ export function getCellComponent(displayMode: TableCellDisplayMode, field: Field
       return DataLinksCell;
     case TableCellDisplayMode.Actions:
       return ActionsCell;
+    case TableCellDisplayMode.Pill:
+      return DefaultCell; // Legacy table doesn't support pill cells, fallback to default
```
Comment (fastfrwrd) [inline packages/grafana-ui/src/components/Table/utils.ts:200]: nice call here.

**Review by fastfrwrd — CHANGES_REQUESTED** [review]

File: packages/grafana-schema/src/common/common.gen.ts (hunk 2 above)
Comment (jesdavpet) [reply packages/grafana-schema/src/common/common.gen.ts:852]: Sounds good, I don't want to start a side-quest into generated code. 👍 Resolving.

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx (hunk 3 above)
Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx]: done in bcd00b3d60f

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx (hunk 4 above)
Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx]: done in bcd00b3d60f

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx (hunk 5 above)
Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx]: done in bcd00b3d60f

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx (hunk 6 above)
Comment (timlevett) [reply packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.tsx]: done in bcd00b3d60f

File: packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        colorMode: 'auto',
+      };
+
+      render(<PillCell {...defaultProps} cellOptions={autoOptions} />);
+      expect(screen.getByText('test-value')).toBeInTheDocument();
+    });
+  });
+});
+
+describe('inferPills', () => {
```
Comment (fastfrwrd) [inline packages/grafana-ui/src/components/Table/TableNG/Cells/PillCell.test.tsx:186]: thank you for this!

**Review by fastfrwrd — APPROVED** [review]

**Comment by marybaldwin** [comment]
Hey @timlevett - I'm prepping for the 12.1 release, do you plan to release this for then? If so, could you submit the [what's new note](https://admin.grafana.com/content-admin/#/collections/whats-new/new) for it?

**Comment by timlevett** [comment]
> What's new draft; https://admin.grafana.com/content-admin/#/collections/whats-new/entries/2025-07-02-pills-for-table-cells

@marybaldwin i drafted this, let me know what you think. we could also add in the video demo i have above or we could do another one.

**Comment by marybaldwin** [comment]
@timlevett - yeah, add the demo and publish!
